astropy/coordinates/baseframe.py

```diff
@@ -1348,19 +1348,20 @@ def _frameattr_equiv(left_fattr, right_fattr):
         before checking for equality in the normal fashion.  This is because
         checking for equality with non-equivalent frames raises an error.
         """
-        if left_fattr is right_fattr:
+        if left is right:
             # shortcut if it's exactly the same object
             return True
-        elif left_fattr is None or right_fattr is None:
+        if left is None or right is None:
             # shortcut if one attribute is unspecified and the other isn't
             return False
 
-        left_is_repr = isinstance(left_fattr, r.BaseRepresentationOrDifferential)
-        right_is_repr = isinstance(right_fattr, r.BaseRepresentationOrDifferential)
-        if left_is_repr and right_is_repr:
+        left_is_repr = isinstance(left, r.BaseRepresentationOrDifferential)
+        if left_is_repr is not isinstance(right, r.BaseRepresentationOrDifferential):
```

Comment:
I find it makes it less clear to have `left_is_repr` but not `right_is_repr`. It might help to just get rid of it altogether, like
```
if isinstance(left, r.BaseRepresentationOrDifferential):
    return (
        isinstance(right, r.BaseRepresentationOrDifferential)
        and  # your complicated all below
```
though to be honest I find
```
if not isinstance(right, r.BaseRepresentationOrDifferential):
    return False

elif isinstance(right_fattr, left_fattr.__class__):
    # if same representation type, compare components.
    return ...

else:
     # convert to cartesian and see if they match
     return ...
```
substantially clearer, despite the ruff rule. To me, just another one we should not enforce...

Alternatively, maybe this is an example that might be nicely done with a `match`?
    ...

Reply:
I agree with @mhvk. This function has a fair number of edge cases. It didn't feel that clean before, and while this reduces the number of return statements, it's still unclear. 

From the comment above... "Primary purpose is to check for equality of representations.  This
        aspect can actually be simplified/removed now that representations have
        equality defined." Perhaps this function is due for a larger refactor?

Reply:
The updated code only needs to check `isinstance(right, r.BaseRepresentationOrDifferential)` at most once so there's no need to store the result in a variable. `isinstance(left, r.BaseRepresentationOrDifferential)` might have to be checked twice, so using a variable is a good idea. I will point out that the last use of `left_is_repr` is no more than three lines below where it was defined, whereas the previous code made use of `left_is_repr` and `right_is_repr` 28 and 27 lines below their definition respectively. I would have thought that this would be the least confusing change I've made.

If you think that the big return block starting at line 1374 is too confusing then I am willing to rewrite that.

Reply:
> From the comment above... "Primary purpose is to check for equality of representations. This aspect can actually be simplified/removed now that representations have equality defined." Perhaps this function is due for a larger refactor?

`BaseRepresentationOrDifferential.__eq__` does exist: https://github.com/astropy/astropy/blob/a2ee8ebd1cfba75ee3872173039deee4795c0179/astropy/coordinates/representation/base.py#L311
However, it seems to be stricter that what this code is doing. It would require `left` and `right` to be instances of the same class, where as currently `right` is only required to be an instance of a subclass of `type(left)`. Furthermore, there's no fallback to a Cartesian representation. That being said, actually using it doesn't cause any tests to fail, so perhaps the differences are not important. It's worth looking into.